usvm-util/src/main/kotlin/org/usvm/collections/immutable/implementations/immutableSet/UPersistentHashSetIterator.kt

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2016-2019 JetBrains s.r.o.
+ * Use of this source code is governed by the Apache 2.0 License that can be found in the LICENSE.txt file.
+ */
+
+package org.usvm.collections.immutable.implementations.immutableSet
+
+internal open class UPersistentHashSetIterator<E>(node: TrieNode<E>) : Iterator<E> {
+    protected val path = mutableListOf(TrieNodeIterator<E>())
+    protected var pathLastIndex = 0
+    private var hasNext = true
+
+    init {
+        path[0].reset(node.buffer)
+        pathLastIndex = 0
+        ensureNextElementIsReady()
+    }
+
+    private fun moveToNextNodeWithData(pathIndex: Int): Int {
+        if (path[pathIndex].hasNextElement()) {
+            return pathIndex
+        }
+        if (path[pathIndex].hasNextNode()) {
+            val node = path[pathIndex].currentNode()
+
+            if (pathIndex + 1 == path.size) {
+                path.add(TrieNodeIterator())
+            }
+            path[pathIndex + 1].reset(node.buffer)
+            return moveToNextNodeWithData(pathIndex + 1)
+        }
+        return -1
+    }
+
+    private fun ensureNextElementIsReady() {
+        if (path[pathLastIndex].hasNextElement()) {
+            return
+        }
+        for (i in pathLastIndex downTo 0) {
+            var result = moveToNextNodeWithData(i)
+
+            if (result == -1 && path[i].hasNextCell()) {
+                path[i].moveToNextCell()
+                result = moveToNextNodeWithData(i)
+            }
+            if (result != -1) {
+                pathLastIndex = result
+                return
+            }
+            if (i > 0) {
+                path[i - 1].moveToNextCell()
+            }
+            path[i].reset(TrieNode.EMPTY.buffer, 0)
+        }
+        hasNext = false
+    }
+
+    override fun hasNext(): Boolean {
+        ensureNextElementIsReady()
```

Comment:
Can you explain why the original kotlinx implementation doesn't invoke `ensureNextElementIsReady` here? Maybe we violate some invariants?